pkg/predicate/predicate.go

```diff
@@ -0,0 +1,174 @@
+// Copyright (c) 2019 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package predicate
+
+import (
+	druidv1alpha1 "github.com/gardener/etcd-druid/api/v1alpha1"
+	"github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/runtime"
+
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/runtime/inject"
+
+	"github.com/gardener/etcd-druid/pkg/common"
+)
+
+// Log is the logger for predicates.
+var (
+	logger = logrus.New()
+)
+
+type or struct {
+	predicates []predicate.Predicate
+}
+
+func (o *or) orRange(f func(predicate.Predicate) bool) bool {
+	for _, p := range o.predicates {
+		if f(p) {
+			return true
+		}
+	}
+	return false
+}
+
+// Create implements Predicate.
+func (o *or) Create(event event.CreateEvent) bool {
+	return o.orRange(func(p predicate.Predicate) bool { return p.Create(event) })
+}
+
+// Delete implements Predicate.
+func (o *or) Delete(event event.DeleteEvent) bool {
+	return o.orRange(func(p predicate.Predicate) bool { return p.Delete(event) })
+}
+
+// Update implements Predicate.
+func (o *or) Update(event event.UpdateEvent) bool {
+	return o.orRange(func(p predicate.Predicate) bool { return p.Update(event) })
+}
+
+// Generic implements Predicate.
+func (o *or) Generic(event event.GenericEvent) bool {
+	return o.orRange(func(p predicate.Predicate) bool { return p.Generic(event) })
+}
+
+// InjectFunc implements Injector.
+func (o *or) InjectFunc(f inject.Func) error {
+	for _, p := range o.predicates {
+		if err := f(p); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// Or builds a logical OR gate of passed predicates.
+func Or(predicates ...predicate.Predicate) predicate.Predicate {
+	return &or{predicates}
+}
+
+// HasOperationAnnotation is a predicate for the operation annotation.
+func HasOperationAnnotation() predicate.Predicate {
+	return FromMapper(MapperFunc(func(e event.GenericEvent) bool {
```

Comment:
Can't you just follow what was done for `LastOperationNotSuccessful` below? Then you don't need all this complicated mapper functions.

```golang
func HasOperationAnnotation() predicate.Predicate {
	f := func(obj runtime.Object) bool {
		etcd, ok := obj.(*druidv1alpha1.Etcd)
		if !ok {
			return false
		}
		return etcd.Annotations[common.GardenerOperation] == common.GardenerOperationReconcile
	}

	return predicate.Funcs{
		CreateFunc: func(event event.CreateEvent) bool {
			return f(event.Object)
		},
		UpdateFunc: func(event event.UpdateEvent) bool {
			return f(event.ObjectNew)
		},
		GenericFunc: func(event event.GenericEvent) bool {
			return f(event.Object)
		},
		DeleteFunc: func(event event.DeleteEvent) bool {
			return f(event.Object)
		},
	}
}
```

Reply:
Makes sense. It would be fine to return true in case of `DeleteEvent` if annotation is not set too, ryt?

Reply:
Hm yes, I guess.